### PR TITLE
[`FA2`] Add flash attention for `GPT-Neo`

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

Adds Flash Attention 2 for `GPT-Neo` as discussed in in this issue - https://github.com/huggingface/transformers/issues/26350 .

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc : @younesbelkada

### CODE REVIEW

**Comment by susnato** [comment]
Again, all tests are passing except the test_flash_attn_2_generate_use_cache tests.
Screenshot from 2023-09-29 14-18-40

Hi @younesbelkada , you might need to run the tests on your end to double check, I am so sorry for the inconvenience.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -50,6 +63,36 @@
 _CHECKPOINT_FOR_DOC = "EleutherAI/gpt-neo-1.3B"
 
 
+# TODO: @younesbelkada move that in pytorch_utils and document it
+if is_torch_fx_available():
```
Comment (younesbelkada) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:67]: I will take care of that in a follow up PR, this seems to be needed to make the torch_fx pass

**Review by younesbelkada — APPROVED** [review]
Very nice speedups ! 

Note that for GPT-Neo it is not possible to train  / run on very long context https://huggingface.co/EleutherAI/gpt-neo-2.7B/blob/main/config.json#L58 as the max position embeddings is limited to 2048 - but this is applicable to all gpt-neo models and not specific to FA-2


Screenshot 2023-10-03 at 13 17 24

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:67]: No no, let's wait until your PR is merged 😉

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 2]
```diff
@@ -253,6 +297,174 @@ def forward(
...
+        if input_dtype == torch.float32:
+            logger.warning_once(
+                "The input hidden states seems to be silently casted in float32, this might be related to"
+                " the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+                " float16."
+            )
+
+            query = query.to(torch.float16)
+            key = key.to(torch.float16)
+            value = value.to(torch.float16)
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:362]: same here let's wait for your PR to be merged

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 3]
```diff
@@ -253,6 +297,174 @@ def forward(
...
+
+        attn_weights_reshaped = attn_output.reshape(bsz, query_length, self.num_heads * self.head_dim)
+        attn_output = self.out_proj(attn_weights_reshaped)
+        attn_output = self.resid_dropout(attn_output)
+
+        outputs = (attn_output, present)
+        if output_attentions:
+            outputs += (attn_weights_reshaped,)
+
+        return outputs  # a, present, (attentions)
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:376]:
```suggestion
        return outputs
```
to remove no?

**Review by ArthurZucker — COMMENTED** [review]
Hi! Same comment as for Llama and Falcon, we ought to add a section on the readme for the expected performance improvements. 
Let's also link the model name to it's doc!

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 3 above)
Comment (susnato) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:376]: yes we should remove it.

**Comment by susnato** [comment]
I have updated the readme.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 1 above)
Comment (susnato) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:67]: Hello @younesbelkada, sorry for disturbing but did you manage to add this method to pytorch_utils?

I would then update the code to make it ready for another review.

**Comment by susnato** [comment]
Hi @younesbelkada, @amyeroberts, I have made the necessary changes and this is ready for another review.

All the `flash attention` tests are passing.

Screenshot from 2023-11-02 19-36-44

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: docs/source/en/model_doc/gpt_neo.md [hunk 4]
```diff
@@ -54,6 +54,47 @@ The `generate()` method can be used to generate text using GPT Neo model.
 >>> gen_text = tokenizer.batch_decode(gen_tokens)[0]
 ```
 
+
+## Combining GPT-Neo and Flash Attention 2
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply docs/source/en/model_doc/gpt_neo.md:58]: +1 on this let me refactor that in another PR

**Comment by susnato** [comment]
Hi @amyeroberts, I have removed all padding_masks.

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 5]
```diff
@@ -322,7 +556,13 @@ def forward(
         head_mask=None,
         use_cache=False,
         output_attentions=False,
+        **kwargs,
     ):
+        if "padding_mask" in kwargs:
```
[the remark this answers is not in the document]
Comment (susnato) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py]: The flash attention for Llama accepts `padding_mask`, so I added it to keep the structure similar :)

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 6]
```diff
@@ -253,6 +288,186 @@ def forward(
...
+        attn_dropout = self.config.attention_dropout if self.training else 0.0
+
+        # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+        # therefore the input hidden states gets silently casted in float32. Hence, we need
+        # cast them back in the correct dtype just to be sure everything works as expected.
+        # This might slowdown training & inference so it is recommended to not cast the LayerNorms
+        # in fp32. (LlamaRMSNorm handles it correctly)
+
+        input_dtype = query.dtype
+        if input_dtype == torch.float32:
```
Comment (amyeroberts) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:346]:
```suggestion 
        if query.dtype == torch.float32:
```

**Review by amyeroberts — APPROVED** [review]
Thanks again for adding this to one of our models! 

There was a recent push of documentation changes to main, so you'll need to rebase to resolve the current conflict.

**Comment by susnato** [comment]
Hi @amyeroberts, the conflict is resolved!

**Comment by susnato** [comment]
I think that we should remove the `padding_mask` from the [startcoder](https://github.com/huggingface/transformers/blob/5964f820db1568d26298b37dea9db328185c7f7c/src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py#L243) too, since it was not an accepted input previously. WDTY?

**Comment by amyeroberts** [comment]
@susnato If you could remove it from there too - that would be great!

**Comment by susnato** [comment]
@amyeroberts , Ok I will open a PR after an hour!
BTW let me know if this PR needs any more changes.

**Pull request merged**